cudax/include/cuda/experimental/__memory_resource/shared_resource.cuh

```diff
@@ -54,42 +51,104 @@ namespace cuda::experimental::mr
-  _LIBCUDACXX_TEMPLATE(class _Resource)
-  _LIBCUDACXX_REQUIRES(
-    __different_resource<_Resource> _LIBCUDACXX_AND(_CUDA_VMR::resource_with<_Resource, _Properties...>))
-  shared_resource(_Resource&& __res)
-      : __resource_(::std::make_shared<::cuda::experimental::mr::any_resource<_Properties...>>(_CUDA_VSTD::move(__res)))
+  static_assert(_CUDA_VMR::resource<_Resource>, "");
+
+  //! @brief Constructs a \c shared_resource refering to an object of type \c _Resource
+  //! that has been constructed with arguments \c __args. The \c _Resource object is
+  //! dynamically allocated with \c new.
+  //! @param __args The arguments to be passed to the \c _Resource constructor.
+  template <class... _Args>
+  explicit shared_resource(_Args&&... __args)
+      : __control_block(new _Control_block{_Resource{_CUDA_VSTD::forward<_Args>(__args)...}, 1})
+  {}
+
+  //! @brief Copy-constructs a \c shared_resource object resulting in an copy that shares
+  //! ownership of the wrapped resource with \c __other.
+  //! @param __other The \c shared_resource object to copy from.
+  shared_resource(const shared_resource& __other) noexcept
+      : __control_block(__other.__control_block)
+  {
+    if (__control_block)
+    {
+      __control_block->__ref_count.fetch_add(1, _CUDA_VSTD::memory_order_relaxed);
+    }
+  }
+
+  //! @brief Move-constructs a \c shared_resource assuming ownership of the resource stored
+  //! in \c __other.
+  //! @param __other The \c shared_resource object to move from.
+  //! @post \c __other is left in a valid but unspecified state.
+  shared_resource(shared_resource&& __other) noexcept
+      : __control_block(_CUDA_VSTD::exchange(__other.__control_block, nullptr))
   {}
 
+  //! @brief Releases the reference held by this \c shared_resource object. If this is the
+  //! last reference to the wrapped resource, the resource is deleted.
+  ~shared_resource()
+  {
+    if (__control_block && __control_block->__ref_count.fetch_sub(1, _CUDA_VSTD::memory_order_acq_rel) == 1)
+    {
+      delete __control_block;
+    }
+  }
+
+  //! @brief Copy-assigns from \c __other. Self-assignment is a no-op. Otherwise, the reference
+  //! held by this \c shared_resource object is released and a new reference is acquired to the
+  //! wrapped resource of \c __other, if any.
+  //! @param __other The \c shared_resource object to copy from.
+  shared_resource& operator=(const shared_resource& __other) noexcept
+  {
+    if (this != &__other)
+    {
+      shared_resource(__other).swap(*this);
+    }
+
+    return *this;
+  }
+
+  //! @brief Move-assigns from \c __other. Self-assignment is a no-op. Otherwise, the reference
+  //! held by this \c shared_resource object is released, while the reference held both \c __other
+  //! is transfered to this object.
+  //! @param __other The \c shared_resource object to move from.
+  /// @post \c __other is left in a valid but unspecified state.
```

Comment:
What does "valid but unspecified state" mean? Is it valid to use other?

Reply:
That is the common phrase of move semantics. 

It means dont touch it